python/openinference-instrumentation/src/openinference/instrumentation/__init__.py

```diff
@@ -12,11 +31,150 @@ def suppress_tracing() -> Iterator[None]:
+        if self._user_id:
+            ctx = set_value(SpanAttributes.USER_ID, self._user_id, ctx)
+        if self._metadata:
+            ctx = set_value(SpanAttributes.METADATA, self._metadata, ctx)
+        if self._tags:
+            ctx = set_value(SpanAttributes.TAG_TAGS, self._tags, ctx)
+
+        self._token = attach(ctx)
+        return
+
+    def detach_context(self) -> None:
+        detach(self._token)
+        return
+
+    def __enter__(self) -> "UsingAttributes":
+        self.attach_context()
+        return self
+
+    async def __aenter__(self) -> "UsingAttributes":
+        self.attach_context()
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback) -> None:
+        self.detach_context()
         return
-    token = attach(set_value(_SUPPRESS_INSTRUMENTATION_KEY, True))
-    yield
-    detach(token)
+
+    async def __aexit__(self, exc_type, exc_value, traceback) -> None:
+        self.detach_context()
+        return
+
+
+class using_session(UsingAttributes):
+    """
+    TBD
+    """
+
+    def __init__(self, session_id: str) -> None:
+        super().__init__(session_id=session_id)
+
+    def __enter__(self) -> "using_session":
+        super().__enter__()
+        return self
+
+    async def __aenter__(self) -> "using_session":
+        super().__aenter__()
+        return self
+
+
+class using_user(UsingAttributes):
+    """
+    TBD
+    """
+
+    def __init__(self, user_id: str) -> None:
+        super().__init__(user_id=user_id)
+
+    def __enter__(self) -> "using_user":
+        super().__enter__()
+        return self
+
+    async def __aenter__(self) -> "using_user":
+        super().__aenter__()
+        return self
+
+
+class using_metadata(UsingAttributes):
+    """
+    TBD
+    """
+
+    def __init__(self, metadata: Dict[str, str]) -> None:
+        super().__init__(metadata=metadata)
+
+    def __enter__(self) -> "using_metadata":
+        super().__enter__()
+        return self
+
+    async def __aenter__(self) -> "using_metadata":
+        super().__aenter__()
+        return self
+
+
+class using_tags(UsingAttributes):
+    """
+    TBD
+    """
+
+    def __init__(self, tags: List[str]) -> None:
+        super().__init__(tags=tags)
+
+    def __enter__(self) -> "using_tags":
+        super().__enter__()
+        return self
+
+    async def __aenter__(self) -> "using_tags":
+        super().__aenter__()
+        return self
+
+
+class using_attributes(UsingAttributes):
+    """
+    TBD
+    """
+
+    def __init__(self, *args: Any, **kwargs: Any) -> None:
+        super().__init__(*args, **kwargs)
+
+    def __enter__(self) -> "using_attributes":
+        super().__enter__()
+        return self
+
+    async def __aenter__(self) -> "using_attributes":
+        await super().__aenter__()
+        return self
+
+
+def get_attributes_from_context() -> Iterator[Tuple[str, AttributeValue]]:
+    for ctx_attr in CONTEXT_ATTRIBUTES:
+        yield ctx_attr, get_value(ctx_attr)
```

Comment:
We should probably only yield if the value is not `None`. Otherwise, won't we wind up sending a bunch of null valued attributes?

Reply:
Good call out. Downstream consumers only append them to the attributes if its not None but its better to do it at creation time as well

Reply:
Done